### PR TITLE
[foxy] reject cancel request if failed to transit to CANCEL_GOAL state. (#791)

### PR DESCRIPTION
**Pull request by jacobperron**
Backport #791 to Foxy.

### CODE REVIEW

**Review by ivanpauno — APPROVED** [review]

**Comment by fujitatomoya** [comment]
I think backport cannot be straight forward to foxy since of Pybind11 modification. I will adjust the code and make sure about that.

File: rclpy/rclpy/action/server.py [hunk 1]
```diff
@@ -28,6 +28,9 @@
 from rclpy.type_support import check_for_type_support
 from rclpy.waitable import NumberOfEntities, Waitable
 
+# Re-export exception defined in _rclpy C extension.
+RCLError = _rclpy.RCLError
+
```
Comment (fujitatomoya) [inline rclpy/rclpy/action/server.py:33]:
```suggestion
```

File: rclpy/rclpy/action/server.py [hunk 2]
```diff
@@ -375,7 +378,15 @@ async def _execute_cancel_request(self, request_header_and_message):
 
             if CancelResponse.ACCEPT == response:
                 # Notify goal handle
-                goal_handle._update_state(GoalEvent.CANCEL_GOAL)
+                try:
+                    # If the goal's just succeeded after user cancel callback
+                    # that will generate an exception from invalid transition.
+                    goal_handle._update_state(GoalEvent.CANCEL_GOAL)
+                except RCLError as ex:
```
Comment (fujitatomoya) [inline rclpy/rclpy/action/server.py:385]: probably we need to use general exception statement here?

```suggestion
                except Exception as ex:
```

**Review by fujitatomoya — COMMENTED** [review]
@jacobperron i do not have access permission to mainline, so left a couple of comments here.

**Comment by jacobperron** [comment]
@fujitatomoya Thanks, I've applied your suggestions.

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14635/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9389/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12300/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14780/)

**Comment by fujitatomoya** [comment]
These seem to be unrelated failures such as `ros2 security generate_policy` failure, retry CI.

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14640/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9393/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12304/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14784/)

**Comment by jacobperron** [comment]
> These seem to be unrelated failures such as ros2 security generate_policy failure

This has been failing for Foxy for a while. I'm fairly sure it is not related to this PR.

I think the Windows failures are all pre-existing too.

**Pull request merged**